lib/src/database/matrix_sdk_database.dart

```diff
@@ -0,0 +1,1590 @@
+    );
+    _roomsBox = _collection.openBox<Map>(
+      _roomsBoxName,
+    );
+    _preloadRoomStateBox = _collection.openBox(
+      _preloadRoomStateBoxName,
+    );
+    _nonPreloadRoomStateBox = _collection.openBox(
+      _nonPreloadRoomStateBoxName,
+    );
+    _roomMembersBox = _collection.openBox(
+      _roomMembersBoxName,
+    );
+    _toDeviceQueueBox = _collection.openBox(
+      _toDeviceQueueBoxName,
+    );
+    _roomAccountDataBox = _collection.openBox(
+      _roomAccountDataBoxName,
+    );
+    _inboundGroupSessionsBox = _collection.openBox(
+      _inboundGroupSessionsBoxName,
+    );
+    _outboundGroupSessionsBox = _collection.openBox(
+      _outboundGroupSessionsBoxName,
+    );
+    _olmSessionsBox = _collection.openBox(
+      _olmSessionsBoxName,
+    );
+    _userDeviceKeysBox = _collection.openBox(
+      _userDeviceKeysBoxName,
+    );
+    _userDeviceKeysOutdatedBox = _collection.openBox(
+      _userDeviceKeysOutdatedBoxName,
+    );
+    _userCrossSigningKeysBox = _collection.openBox(
+      _userCrossSigningKeysBoxName,
+    );
+    _ssssCacheBox = _collection.openBox(
+      _ssssCacheBoxName,
+    );
+    _presencesBox = _collection.openBox(
+      _presencesBoxName,
+    );
+    _timelineFragmentsBox = _collection.openBox(
+      _timelineFragmentsBoxName,
+    );
+    _eventsBox = _collection.openBox(
+      _eventsBoxName,
+    );
+    _seenDeviceIdsBox = _collection.openBox(
+      _seenDeviceIdsBoxName,
+    );
+    _seenDeviceKeysBox = _collection.openBox(
+      _seenDeviceKeysBoxName,
+    );
+
+    // Check version and check if we need a migration
+    final currentVersion = int.tryParse(await _clientBox.get('version') ?? '');
+    if (currentVersion == null) {
+      await _clientBox.put('version', version.toString());
+    } else if (currentVersion != version) {
+      await _migrateFromVersion(currentVersion);
+    }
+
+    return;
+  }
+
+  Future<void> _migrateFromVersion(int currentVersion) async {
+    Logs().i('Migrate store database from version $currentVersion to $version');
+    await clearCache();
+    await _clientBox.put('version', version.toString());
+  }
+
+  @override
+  Future<void> clear() => _collection.clear();
+
+  @override
+  Future<void> clearCache() => transaction(() async {
+        await _roomsBox.clear();
+        await _accountDataBox.clear();
+        await _preloadRoomStateBox.clear();
+        await _nonPreloadRoomStateBox.clear();
+        await _roomMembersBox.clear();
+        await _eventsBox.clear();
+        await _timelineFragmentsBox.clear();
+        await _outboundGroupSessionsBox.clear();
+        await _presencesBox.clear();
+        await _clientBox.delete('prev_batch');
+      });
+
+  @override
+  Future<void> clearSSSSCache() => _ssssCacheBox.clear();
+
+  @override
+  Future<void> close() async => _collection.close();
+
+  @override
+  Future<void> deleteFromToDeviceQueue(int id) async {
+    await _toDeviceQueueBox.delete(id.toString());
```

Comment:
This probably doesn't matter from a performance perspective, but still feels bad :D

How bad would it be to type the key? We could still toString it in the box, but that would also make some of the documentation more obvious, what is a TupleKey and what isn't?

(Nitpick)

Reply:
Hmm why making the key type-able when we internally convert it toString anyway? The TupleKey also is just outside of the key/value abstraction layer and not different from how it was done with hive.

I would more suggest that we try to implement something like additional multiple indexes for a column in the future which then could also be typed. This can be mapped quite good to SQL and indexeddb and could make some things much more performant and elegant in the future. however I would still suggest that we start with a drop-in replacement for Hive for now

Reply:
But if we add the multiple indices at a later point, that would require a migration, right? I guess that isn't too bad, it would just be a bit weird, since it would be a level lower than I expected :D

Feel free to resolve this by writing an issue for a possible improvement!

Reply:
https://github.com/famedly/matrix-dart-sdk/issues/1634